massa-execution-worker/src/speculative_ledger.rs

```diff
@@ -13,23 +13,21 @@ use massa_models::{Address, Amount};
 use parking_lot::RwLock;
 use std::sync::Arc;
 
+use crate::active_history::{ActiveHistory, HistorySearchResult};
+
 /// The `SpeculativeLedger` contains an thread-safe shared reference to the final ledger (read-only),
 /// a list of existing changes that happened o the ledger since its finality,
 /// as well as an extra list of "added" changes.
 /// The `SpeculativeLedger` makes it possible to transparently manipulate a virtual ledger
 /// that takes into account all those ledger changes and allows adding more
 /// while keeping track of all the newly added changes, and never writing in the final ledger.
-pub struct SpeculativeLedger {
+pub(crate) struct SpeculativeLedger {
     /// Thread-safe shared access to the final state. For reading only.
     final_state: Arc<RwLock<FinalState>>,
 
-    /// Accumulation of changes that previously happened to the ledger since finality.
-    /// This value is not modified by changes applied to the `SpeculativeLedger`.
-    ///
-    /// TODO maybe have the history directly here,
-    /// so that we can avoid accumulating all the changes at every slot
-    /// but only lazily query addresses backwards in history (to avoid useless computations) with caching
-    previous_changes: LedgerChanges,
+    /// History of the outputs of recently executed slots.
+    /// Slots should be consecutive, newest at the back.
+    active_history: Arc<RwLock<ActiveHistory>>,
```

Comment:
why arc rwlock ? it was not necessary before this PR

Reply:
See https://github.com/massalabs/massa/pull/2701#discussion_r901504632